pulsar-io/nifi/src/main/java/org/apache/pulsar/io/nifi/NiFiSource.java

```diff
@@ -0,0 +1,188 @@
+        this.isRunning = false;
+        if (null != client) {
+            client.close();
+        }
+
+        if (null != runnerThread) {
+            runnerThread.interrupt();
+            runnerThread.join();
+            runnerThread = null;
+        }
+    }
+
+    protected void start() throws Exception {
+        runnerThread = new Thread(new ReceiveRunnable());
+        runnerThread.setName("Nifi Source Thread");
+        runnerThread.start();
+    }
+
+    class ReceiveRunnable implements Runnable {
+
+        public ReceiveRunnable() {
+
+        }
+
+        @Override
+        public void run() {
+            Transaction transaction = null;
+            while (isRunning) {
+                try {
+                    transaction = client.createTransaction(TransferDirection.RECEIVE);
+                } catch (IOException ioe) {
+                    log.warn("Created NiFi transaction Failed", ioe);
+                }
+
+                if (null == transaction) {
+                    try {
+                        Thread.sleep(waitTimeMs);
+                    } catch (InterruptedException ioe) {
+                        log.warn("transaction could not be created, waiting and will try again " + waitTimeMs + " milliseconds.");
+                    }
+                    continue;
+                }
+
+                try {
+                    DataPacket dataPacket = transaction.receive();
+                    if (null == dataPacket) {
+                        // no data available. Wait a bit and try again
+                        try {
+                            Thread.sleep(waitTimeMs);
+                        } catch (InterruptedException ioe) {
+                            log.warn("dataPacket could not be received, waiting and will try again " + waitTimeMs + " milliseconds.");
+                        }
+                        continue;
+                    }
+
+                    final List<NiFiDataPacket> dataPackets = Lists.newArrayList();
+                    do {
+                        // Read the data into a byte array and wrap it along with the attributes into a NiFiDataPacket.
+                        final InputStream inStream = dataPacket.getData();
+                        final byte[] data = new byte[(int) dataPacket.getSize()];
+                        StreamUtils.fillBuffer(inStream, data);
+
+                        final Map<String, String> attributes = dataPacket.getAttributes();
+                        final NiFiDataPacket NiFiDataPacket = new StandardNiFiDataPacket(data, attributes);
+                        dataPackets.add(NiFiDataPacket);
+                        dataPacket = transaction.receive();
+                    } while (dataPacket != null);
+
+                    for (NiFiDataPacket dp : dataPackets) {
+                        consume(new NiFiRecord(dp));
+                    }
+
+                    // Confirm transaction to verify the data
+                    transaction.confirm();
```

Comment:
I don't think you should confirm/complete the transaction here. confirm/complete or error should be done by the NiFiRecord#ack() and NiFiRecord#fail()